src/main/kotlin/com/cognifide/gradle/aem/instance/tasks/InstanceRestore.kt

```diff
@@ -0,0 +1,121 @@
+package com.cognifide.gradle.aem.instance.tasks
+
+import com.cognifide.gradle.aem.common.AemException
+import com.cognifide.gradle.aem.common.AemTask
+import com.cognifide.gradle.aem.common.Formats
+import com.cognifide.gradle.aem.common.file.FileOperations
+import com.cognifide.gradle.aem.common.file.resolver.FileResolver
+import com.cognifide.gradle.aem.common.onEachApply
+import com.cognifide.gradle.aem.common.tasks.LocalInstanceTask
+import com.cognifide.gradle.aem.instance.InstanceException
+import com.cognifide.gradle.aem.instance.LocalInstance
+import com.cognifide.gradle.aem.instance.names
+import com.fasterxml.jackson.annotation.JsonIgnore
+import java.io.File
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.TaskAction
+
+open class InstanceRestore : LocalInstanceTask() {
+
+    init {
+        description = "Restores AEM instance(s) from backup file."
+    }
+
+    private val downloadDir = AemTask.temporaryDir(aem.project, TEMPORARY_DIR)
+
+    private val fileResolver = FileResolver(aem, downloadDir).apply { group(GROUP_EXTRA) {} }
+
+    @Internal
+    val uploadUrl = aem.props.string("backup.uploadUrl")
+
+    @Internal
+    val backupUrl = aem.props.string("backup.downloadUrl")
+
+    /**
+     * Defines backup selection rule.
+     *
+     * By default takes desired backup by name (if provided) or takes most recent backup
+     * (file names sorted lexically / descending).
+     */
+    @JsonIgnore
+    var backupSelector: Collection<String>.() -> String? = {
+        val backupName = aem.props.string("backup.name") ?: ""
+        when {
+            backupName.isNotBlank() -> firstOrNull { it == backupName }
+            else -> sortedByDescending { it }.firstOrNull()
+        }
+    }
+
+    @TaskAction
+    fun restore() {
+        val uncreatedInstances = instances.filter { !it.created }
+        if (uncreatedInstances.isEmpty()) {
+            logger.info("No instance(s) to restore")
+            return
+        }
+
+        logger.info("Restoring instances: ${uncreatedInstances.names}")
+        restoreFromBackup(uncreatedInstances, selectBackup())
+
+        val restoredInstances = uncreatedInstances.filter { it.created }
+        aem.notifier.notify("Instance(s) restored", "Which: ${restoredInstances.names}")
+    }
+
+    private fun selectBackup() = when {
+        backupUrl != null -> fileResolver.run { backupUrl.run { url(this) } }.file
+        uploadUrl != null -> {
+            val name = backupSelector(aem.fileTransfer(uploadUrl).list())
+                    ?: throw AemException("No backups to restore. Please perform backup before restoring.")
+            fileResolver.run { backupUrl(uploadUrl, name).run { url(this) } }.file
+        }
+        aem.tasks.named<InstanceBackup>(InstanceBackup.NAME).get().available.isNotEmpty() -> {
```

Comment:
introduce var for `aem.tasks.named<InstanceBackup>(InstanceBackup.NAME).get().available`

Reply:
OK